### PR TITLE
Ffdpacking

### LINKED ISSUE #934: Add a new packing algorithm based on binpacking

### PR DESCRIPTION
**Pull request by avflor**
fixes #934 

The branch contains a new packing algorithm based on binpacking that can support heterogeneous containers and can be used with schedulers like YARN.

### CODE REVIEW

**Comment by objmagic** [comment]
Thx. Please sign the [CLA](https://engineering.twitter.com/opensource/cla) first.

**Comment by avflor** [comment]
I see there are some checkstyle errors which I don't get when building locally and running the tests. Is there any way to run chcekstyle alone and figure out what is wrong?

**Comment by maosongfu** [comment]
@avflor 
Try with: bazel build --config=darwin --experimental_action_listener=tools/java:compile_java heron/...

**Comment by avflor** [comment]
@maosongfu Thanks! I tried this command and it still works locally. I'm working on Ubuntu but we also tried on a mac and we don't get any checkstyle errors there as well.

**Comment by objmagic** [comment]
I checked out your branch and ran

``bazel build --config=darwin --experimental_action_listener=tools/java:compile_java heron/packing/...``

I can see lots of style errors.

**Comment by maosongfu** [comment]
@avflor Maybe you need "bazel clean" first? @billonahill may have better ideas on this.

**Comment by ashvina** [comment]
Yes. I can see the errors after "bazel clean".

**Comment by kramasamy** [comment]
@objmagic - can you cut and paste the checkstyle errors for @avflor?

**Comment by objmagic** [comment]
@ashvina are you the same guy as @avflor? and who is @avrilia? what is going on here?

**Comment by kramasamy** [comment]
@avflor - one broader comment is to write a documentation about this packing algorithm.

- What does the algorithm do?
- When is it useful and its pros and cons?
- How to enable it?

**Comment by avflor** [comment]
@kramasamy I fixed the checkstyle errors. After bazel clean I was able to see them and I checked in the new code. I will write some extra documentation. I have some comments about the algo in the code but I can expand.

@objmagic Thanks for noticing that. There has been some weird configuration issue and apparently GitHub linked my email with an old GitHub account (avrilia). I will fix this and from now on only avflor will be used.

**Comment by objmagic** [comment]
@avflor please fix two Java Docs error [here](https://travis-ci.org/twitter/heron/builds/144013684#L1665). You should use ``&lt;`` and ``&gt;``.

**Comment by kramasamy** [comment]
@ajorgensen - can you review this PR since you are the closest to write a new packing algorithm?

File: heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+ * ((# containers used by FFD)* size of each container
+ * + 2 * (size of a container that includes one instance with default resource requirements)).
+ * The extra 2 is considered for Heron internal container (i.e. the one containing Scheduler
+ * and TMaster) and for the YARN application master,
+ * <p>
+ * 7. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of ram for an instance is less than the minimal required value.
+ */
+
+public class FFDPacking implements IPacking {
```
Comment (ajorgensen) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: `FFD` is a pretty esoteric term. What if we expanded it to `FirstFitDecreasing` or `BinPacking`? I feel like expanding out the name would make it more discoverable.

File: heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+      containerPlanMap.put(containerId, containerPlan);
+      topologyRam += containerRam;
+      topologyCpu += containerCpu;
+      topologyDisk += containerDiskInBytes;
+    }
+
+    // Take the heron internal container into account and the application master for YARN
+    // scheduler
+    topologyRam += 2 * (instanceRamDefault + DEFAULT_RAM_PADDING_PER_CONTAINER);
```
Comment (ajorgensen) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: So this calculation is not generalizable to other schedulers correct? For example Aurora only has 1 extra container, I am not familiar with YARN but it looks like in this case there are 2 extra containers correct?.

Comment (ashvina) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: Good catch. Each topology running on YARN needs one extra container for application master. However the resources allocated to AM container and TMaster container are constant and independent of instanceRam and default padding. Which means the total resources allocated to a topology is a `constant + cost-of-deploying-instances` for both `yarn` and `aurora` clusters.

File: heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java (hunk 1 above)
Comment (avflor) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: Sure, I can do that. Thanks!

File: heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java (hunk 2 above)
Comment (ajorgensen) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: I've been thinking about this a little more recently and would be curious what your thoughts are about generic schedulers for all platforms and whether or not we should start splitting some of the at least shell implementations up into separate scheduling technologies (yarn, aurora, mesos, etc). I think that trying to generalize the requirements of all of those platforms into one algorithm might actually be making this more complex than it needs to be and certainly some schedulers will have different trade-offs and be able to make certain optimizations based on how they work. I dont necessarily think we need to do that in this case but I think its worth thinking about how that would work and whether the level effort/complexity that splitting them out into separate scheduler implementations might introduce. @kramasamy I'd be curious of your thoughts as well.

Comment (avflor) [inline heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java]: @ajorgensen We had a discussion with @ashvina about the same topic. The main limitation now is the fact that some schedulers support heterogeneous containers whereas others do not. However, the algorithm should be able to operate on both types of schedulers. The algorithm mainly decides how to pack instances in containers and should provide this packing to the scheduler. The scheduler can then decide to use homogeneous sor heterogeneous containers to represent this packing. I think @ashvina will open an issue on that. In this way, we would be able to use RoundRobin and binpacking on both Aurora and Yarn.

**Comment by kramasamy** [comment]
@avflor @ajorgensen @ashvina - I think the consensus to keep the packing algorithms and schedulers orthogonal - the PR for this is here

https://github.com/twitter/heron/pull/1079

**Comment by kramasamy** [comment]
Is FFD packing algorithm is good enough to merge?

**Comment by avflor** [comment]
I made some updates to address some comments and also changed some configuration parameters. I will check it in tomorrow.

**Comment by avflor** [comment]
@ajorgensen Let me know if you have other comments and I will address them tomorrow before I push the current version.

**Comment by ajorgensen** [comment]
@avflor I think aside from my initial comments it looks good to me. I have other aspirations to have those default padding values be configurable as well but I don't believe that needs to happen in this pass.

**Comment by kramasamy** [comment]
Can we open an issue for tracking?

**Comment by ajorgensen** [comment]
Opened one here: https://github.com/twitter/heron/issues/1084

**Comment by avflor** [comment]
@ajorgensen @kramasamy @maosongfu I submitted a revised version of the binpacking algorithm. I noticed when I run in local mode that the number of stream managers is picked based on the topology config. However, the binpacking algorithm decides on its own how many containers to use. The number of stream managers launched should be based on the result of packing algorithm. I enhanced the SchedulerMain.java code to take the value from the PackedPlan. I made the necessary change to the junit tests as well. I also changed the build files so that the Heron client can use the binpacking algorithm. Finally, I made the padding configurable and changed the binpacking algorithm to accommodate it. Please let me know if you have any further comments.

**Comment by kramasamy** [comment]
@avflor - excellent. @ajorgensen @maosongfu - if you are fine with the PR, we can merge it.

@avflor - should we check in the documentation as a part of this PR as we discussed earlier or it will go in a separate PR?

**Comment by avflor** [comment]
@kramasamy Karthik, what do you think about me submitting the documentation PR as a separate one next week?  I need some more time to look into the tools and the setup. In the meantime, I have comments on how to use the algorithm in the FirstFitDecreasing.java class.

**Comment by kramasamy** [comment]
Ok. Please create a github issue and reference it here.

**Comment by avflor** [comment]
First fit decreasing documentation: #1098

**Comment by ajorgensen** [comment]
PR looks good to me!

**Comment by maosongfu** [comment]
looks good to me. 👍

**Pull request merged**